### ISSUE TITLE
Make [UInt8] to Base64 String public

### ISSUE THREAD

**Issue by fabianfett** [apple/swift-nio#1265]
### Expected behavior

TLDR: The `String.init(base64Encoding array: Array<UInt8>)` should be public.

There is a quite simple base64 implementation in `NIOWebsocket`. I was wondering if it would make sense to move this somewhere, where it could be accessed from other packages as well? `NIOWebSocket` probably wouldn't be the best package for that, I guess.
https://github.com/apple/swift-nio/blob/master/Sources/NIOWebSocket/Base64.swift

I am well aware of the comment: 

> This function performs the world's most naive Base64 encoding: no attempts to use a larger lookup table or anything intelligent like that, just shifts and masks. This works fine, for now: the purpose of this encoding is to avoid round-tripping through Data, and the perf gain from avoiding that is more than enough to outweigh the silliness of this code.

But I guess making this public would maybe encourage more people to use it and don't roll their own. And the perf gain from not using `Data` should be applicable for any consumer of Swift-NIO.

### Actual behavior
The String initializer to create a base64 string from [UInt8] is internal.

### Steps to reproduce

https://github.com/apple/swift-nio/blob/5a7b23d8b5b54349b263c0676a8d47b10d116e8a/Sources/NIOWebSocket/Base64.swift#L34

### If possible, minimal yet complete reproducer code (or URL to code)

Not applicable.

### SwiftNIO version/commit hash

v2.10.1

### Swift & OS version (output of `swift --version && uname -a`)

Not applicable.

**Comment by Lukasa** [comment]
Thanks for this issue!

I think in general I don't want swift-nio to vend small utilities like this. I agree that having a general base-64 encode and decode function is useful, but swift-nio is not a very good place to put such a function. Given that swift-nio is permissively licensed, I would recommend taking the following steps:

1. Creating a new swift package containing the contents of that file, making the initializer public, and announcing the package on the Swift forums.
2. Proposing the code for inclusion in the standard library preview package, possibly by making it more generic (i.e. by using `Collection` instead of `Array`).
3. Eventually aim to get the code in the standard library.

How does that sound?

**Comment by karwa** [comment]
Hey, I just came across this issue via your package announcement on Reddit. As it happens, I also had reason to investigate this a few weeks ago (swiftplot communicates with jupyter via base64-encoded PNG images).

Foundation has some utilities to encode/decode base64, but the API is fairly restrictive as it is a direct translation of the Obj-C interface (which only deals in String and Data) to Swift. Luckily, it turns out to be [extremely easy](https://gist.github.com/karwa/c859111036da4e809e95a0d1ef099e54) to make it generic. It was more of a proof-of-concept so I didn't bother to really consider the shape of the public API or optimise it, but there are some copies which could trivially be removed (e.g. the encoding/decoding result is always an Array which gets copied in to a String/Data object; we could avoid that by passing an `inout Collection` and appending the bytes to it).

However, Foundation does not follow an open evolution process, so we'd need Apple's Foundation team to agree to making those changes to the interface (CC @parkera). If they do, we could focus on optimising that single implementation and all Swift projects on all platforms would be able to take advantage of it.

**Comment by fabianfett** [comment]
Hi @karwa,
please see this issue for further information and discussion: https://github.com/apple/swift-nio/issues/1322
